holoviews/tests/operation/test_operation.py

```diff
@@ -126,13 +232,71 @@ def test_qmesh_curvilinear_edges_contours(self):
+                                      [0.08333333, 0.0], [0.0, -0.08333333]])]]]
+        np.testing.assert_array_almost_equal(op_contours.holes(), expected_holes)
+
+    def test_image_contours_filled_multi_holes(self):
+        img = Image(np.array([[0, 0, 0, 0, 0], [0, 1, 0, 1, 0], [0, 0, 0, 0, 0]]))
+        # Single polygon with two holes
+        op_contours = contours(img, filled=True, levels=[-0.5, 0.5])
+        data = [[(-0.4, -0.3333333, 0), (-0.2, -0.3333333, 0), (0, -0.3333333, 0),
+                 (0.2, -0.3333333, 0), (0.4, -0.3333333, 0), (0.4, 0, 0), (0.4, 0.3333333, 0),
+                 (0.2, 0.3333333, 0), (0, 0.3333333, 0), (-0.2, 0.3333333, 0), (-0.4, 0.3333333, 0),
+                 (-0.4, 0, 0), (-0.4, -0.3333333, 0)]]
+        polys = Polygons(data, vdims=img.vdims[0].clone(range=(-0.5, 0.5)))
+        self.assertEqual(op_contours, polys)
+        expected_holes = [[[np.array([[-0.2, -0.16666667], [-0.3, 0], [-0.2, 0.16666667], [-0.1, 0],
+                                      [-0.2, -0.16666667]]),
+                            np.array([[0.2, -0.16666667], [0.1, 0], [0.2, 0.16666667], [0.3, 0],
+                                      [0.2, -0.16666667]])]]]
+        np.testing.assert_array_almost_equal(op_contours.holes(), expected_holes)
+
+    def test_image_contours_filled_empty(self):
         img = Image(np.array([[0, 1, 0], [3, 4, 5.], [6, 7, 8]]))
-        op_contours = contours(img, filled=True, levels=[2, 2.5])
-        data = [[(0., 0.166667, 2.25), (0.333333, 0.166667, 2.25), (0.333333, 0.2, 2.25), (0., 0.222222, 2.25),
-                 (-0.333333, 0.111111, 2.25), (-0.333333, 0.055556, 2.25), (0., 0.166667, 2.25)]]
-        polys = Polygons(data, vdims=img.vdims[0].clone(range=(2, 2.5)))
+        # Contour level outside of data limits
+        op_contours = contours(img, filled=True, levels=[20.0, 23.0])
+        polys = Polygons([[]], vdims=img.vdims[0].clone(range=(20.0, 23.0)))
         self.assertEqual(op_contours, polys)
 
+    def test_image_contours_filled_auto_levels(self):
+        z = np.array([[0, 1, 0], [3, 4, 5.], [6, 7, 8]])
+        img = Image(z)
+        for nlevels in range(3, 20):
+            op_contours = contours(img, filled=True, levels=nlevels)
+            levels = [item['z'] for item in op_contours.data]
+            print(nlevels, len(levels), levels)
+            assert len(levels) <= nlevels + 1
+            delta = 0.5*(levels[1] - levels[0])
+            assert np.min(levels) <= z.min() + delta
+            assert np.max(levels) >= z.max() - delta
+
+    def test_image_contours_filled_x_datetime(self):
+        x = np.array(['2023-09-01', '2023-09-05', '2023-09-09'], dtype='datetime64')
+        y = np.array([6, 7])
+        z = np.array([[0, 2, 0], [0, 2, 0]])
+        img = Image((x, y, z))
+        msg = 'Datetime spatial coordinates are not supported for filled contour calculations.'
```

Comment:
Be careful with `.` as it is treated as any character by `pytest.raises` match is treated as regex. It does not matter here, though.

Reply:
I'll make it a raw string.